packages/wrangler/src/__tests__/tail.test.ts

```diff
@@ -0,0 +1,303 @@
+import { mockAccountId, mockApiToken } from "./helpers/mock-account-id";
+import { setMockResponse } from "./helpers/mock-cfetch";
+import { mockConsoleMethods } from "./helpers/mock-console";
+import { WS } from "./helpers/mock-web-socket";
+import { runInTempDir } from "./helpers/run-in-tmp";
+import { runWrangler } from "./helpers/run-wrangler";
+import type Websocket from "ws";
+
+// change this if you're testing using debug mode
+// (sends all logs regardless of filters)
+const DEBUG = false;
+
+function serialize(message: unknown): Websocket.RawData {
+  return Buffer.from(JSON.stringify(message), "utf-8");
+}
+
+function deserializeToJson(message: Websocket.RawData): string {
+  return JSON.stringify(JSON.parse(message.toString()), null, 2);
+}
+
+// This function should be called at the start of each test.
+// You don't have to use anything it returns, but if we don't
+// mock out these endpoints then the test just hangs.
+// Since we do need to access the properties returned by this
+// function, we can't stick it in a beforeEach (i don't think)
+function setupWebsocketMocks(websocketURL = "ws://localhost:1234") {
+  const server = new WS(websocketURL);
+  const tailCreationRequests = mockCreateTailRequest(websocketURL);
+  const tailDeletionRequests = mockDeleteTailRequest();
+
+  return { server, tailCreationRequests, tailDeletionRequests };
+}
+
+describe("tail", () => {
+  runInTempDir();
+  mockAccountId();
+  mockApiToken();
+
+  // clean up any outstanding connections
+  afterEach(() => WS.clean());
+
+  const std = mockConsoleMethods();
+
+  /* API related functionality */
+
+  it("creates and then delete tails", async () => {
+    const { server, tailCreationRequests, tailDeletionRequests } =
+      setupWebsocketMocks();
+
+    try {
+      expect(tailCreationRequests.count).toStrictEqual(0);
+
+      await runWrangler("tail test-worker");
+
+      await expect(server.connected).resolves.toBeTruthy();
+      expect(tailCreationRequests.count).toStrictEqual(1);
+      expect(tailDeletionRequests.count).toStrictEqual(0);
+    } finally {
+      server.close();
+      expect(tailDeletionRequests.count).toStrictEqual(1);
+    }
+  });
+
+  /* filtering */
+
+  it("sends sampling rate filters", async () => {
+    const { server } = setupWebsocketMocks();
```

Comment:
every test has a setupWebSocketMocks() call and a finally clause that closes it. Maybe it should go to into a beforeEach/afterEach? that way you can avoid the try/finally as well.